docs/best-practices/gitops-with-kruise.md

```diff
@@ -0,0 +1,183 @@
+- Install argo-cd, Reference [Official Documents](https://argo-cd.readthedocs.io/en/stable/getting_started/)
+- Install OpenKruise，Reference [Official Documents](https://openkruise.io/docs/installation/)
+
+### Declarative Applications in Git Repository
+1. OpenKruise provides a guestbook application [demo](https://github.com/openkruise/samples).
+It consists of guestbook, redis, through openKruise CloneSet to complete the deployment of the application. As follows:
+
+![guestbook](/img/docs/best-practices/guestbook.png)
+
+```yaml
+## CloneSet & Service
+apiVersion: apps.kruise.io/v1alpha1
+kind: CloneSet
+...
+spec:
+  replicas: 3
+  ...
+  template:
+  spec:
+    containers:
+    - name: guestbook
+      image: 'k8s.gcr.io/guestbook:v3'
+      env:
+      - name: "VERSION"
+        value: "v1"
+      ports:
+      - name: http-server
+        containerPort: 3000
+---
+kind: Service
+apiVersion: v1
+...
+spec:
+  ports:
+  - port: 3000
+    targetPort: http-server
+  selector:
+    app: guestbook
+    type: LoadBalancer
+
+### redis master/slave is also similar configuration
+apiVersion: apps.kruise.io/v1alpha1
+kind: CloneSet
+...
+spec:
+  replicas: 1
+  ...
+  template:
+    spec:
+    containers:
+    - name: redis-master
+      image: 'k8s.gcr.io/redis:e2e'
+      ports:
+      - name: redis-server
+        containerPort: 6379
+---
+kind: Service
+...
+spec:
+  ports:
+  - port: 6379
+    targetPort: redis-server
+  selector:
+    app: redis
+    role: master
+```
+
+2. For a variety of environments such as test, pre, and prd, different branches can be used to represent them.
+For example, there exists **dev_branch** and **master** corresponding to the git configuration for test and prd environments, as follows:
+
+![git_envs](/img/docs/best-practices/git_envs.png)
+
+### Configuring Argo Application, And Deploying Guestbook
+1. Argo-cd provides a [Declarative Setup](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/), as follows:
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  ## Represents dev env deployment configuration
+  name: guestbook-dev
+  namespace: argocd
+spec:
+  project: default
+  source:
+    repoURL: https://github.com/zmberg/samples.git
+    ## dev branch
+    targetRevision: dev_branch
+    path: guestbook
+  destination:
+    server: https://kubernetes.default.svc
+    ## deploy in default namespace
+    namespace: default
+```
+
+2. Create the Application CRD resource into k8s cluster, and complete the deployment using the argocd command line tool,
+refer to the [documentation](https://argo-cd.readthedocs.io/en/stable/getting_started/), as follows:
+
+![guestbook](/img/docs/best-practices/argo_sync_unknown.png)
+
+**Note: After publishing through argo-cd, we found a very strange problem. argo-cd actually did not determine whether the cloneSet publishing is complete and successful,
```

Comment:
in best practice document, we should be describe the solution, not the process we found the solution. So we should delete narrative sentence such as  "we found a very strange problem"